[http] Headermap Lazy map addition

adisuissa opened the pull request:
Commit Message: Adding a lazy map to the current header-map implementation.

Additional Description: This is a continuation of #9261. Adding a map to headermap's HeaderList to access headers by key in O(1) instead of O(n).
In addition some benchmarks that emulate encoding and decoding of headers, and route matching by headers are added.

Risk Level: High - changes to headermap.
Testing: Added `RemoveIf` test, and some perf tests.
Docs Changes: Once the design is settled, I'll update the [headermap doc](https://github.com/envoyproxy/envoy/blob/master/source/docs/header_map.md)
Release Notes: None.

Signed-off-by: Adi Suissa-Peleg <adip@google.com>

adisuissa commented:
Newly added benchmarks description:
- _headerMapImplEmulateH1toH2Upgrade_ - emulates the _encodeHeaders_ (H2) code, when upgrading a request from H1 to H2.
- _headerMapImplEmulateH2toH1Upgrade_ - emulates the _decodeHeaders_ (H2) code, when upgrading a response from H2 to H1.
- _manyCountryRoutesLongHeaders_ - emulates the route matching code where there are 250 backends, and the routing is done using a header ("x-country"). A request with ~100 headers is dispatched, and its requested country header doesn't exist in any route. This causes the old headermap implementation to iterate over all header keys, and many comparisons of the "x-country" header.

Headermap benchmark results (average of 3 executions):

Benchmark name | No lazy map | threshold=0 | threshold=1 | threshold=2 | threshold=3 | threshold=4 | threshold=5 | threshold=6 | threshold=7 | threshold=8 | threshold=9 | threshold=10
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
headerMapImplCreate | 13.3333 | 16.0333 | 15.3667 | 15.3 | 15.9667 | 15.8 | 15.4 | 15.4333 | 15.3667 | 16.3 | 15.4333 | 16.3
headerMapImplSetReference/0 | 59.1 | 73.7667 | 78.0333 | 45.6667 | 45.9333 | 46.6667 | 45.7333 | 45.9667 | 46.1 | 47.7333 | 45.8667 | 47
headerMapImplSetReference/1 | 68.6 | 74.2 | 80.0667 | 74.0667 | 50.4333 | 50.6333 | 50.2 | 50.6333 | 50.5667 | 51.9333 | 50.4667 | 53.5667
headerMapImplSetReference/5 | 105 | 73.6 | 77.8333 | 74.0333 | 75.5667 | 75.8 | 74.1333 | 74.9667 | 66.2667 | 67 | 66.1 | 70.4333
headerMapImplSetReference/10 | 158 | 73.4 | 86.4333 | 79.8 | 82.5 | 84.2667 | 82.6667 | 83.4667 | 85.4667 | 82.7333 | 85.4 | 85.1333
headerMapImplSetReference/50 | 355.667 | 73.5667 | 87 | 87.2 | 80.6667 | 79.6333 | 77.3667 | 79.3333 | 79.4667 | 81.9 | 81.4333 | 84.4
headerMapImplGet/0 | 12.6667 | 19.0667 | 19.6333 | 14.2667 | 14.1333 | 14.1 | 13.9333 | 14.3667 | 14.3333 | 14.4333 | 14.3 | 14.1333
headerMapImplGet/1 | 19.9667 | 19.0667 | 19.6333 | 19.6667 | 20.6667 | 20.7 | 20.7333 | 20.8333 | 20.6667 | 20.7 | 20.8 | 20.7
headerMapImplGet/5 | 48.6333 | 19.0667 | 19.6333 | 19.6667 | 19.6667 | 19.6333 | 19.6333 | 19.8333 | 46.0333 | 46.1333 | 46.0667 | 46.0667
headerMapImplGet/10 | 89.4333 | 19.1 | 20.9667 | 19.6667 | 21.5 | 21.6333 | 21.6333 | 21.8 | 21.6 | 20.6667 | 22.6667 | 23.2667
headerMapImplGet/50 | 276.333 | 19.0667 | 22.3 | 23.5667 | 21.6 | 21.2 | 20 | 21.3667 | 19.8 | 21.8667 | 21.8333 | 25.5
headerMapImplGetInline/0 | 5.12333 | 5.1 | 5.30667 | 5.10667 | 5.11 | 5.10667 | 5.10333 | 5.40667 | 5.10667 | 5.10333 | 5.27667 | 5.11
headerMapImplGetInline/1 | 5.12667 | 5.06667 | 5.11 | 5.10667 | 5.11 | 5.10667 | 5.10667 | 5.34 | 5.10333 | 5.10667 | 5.12667 | 5.10667
headerMapImplGetInline/5 | 5.12667 | 5.08 | 5.11333 | 5.11 | 5.11667 | 5.12333 | 5.10333 | 5.33333 | 5.11333 | 5.10333 | 5.11 | 5.11333
headerMapImplGetInline/10 | 5.12 | 5.07333 | 5.10333 | 5.11667 | 5.10667 | 5.11333 | 5.10333 | 5.33667 | 5.12333 | 5.11 | 5.11333 | 5.11667
headerMapImplGetInline/50 | 5.12667 | 5.06 | 5.30333 | 5.10333 | 5.10667 | 5.11 | 5.11333 | 5.30333 | 5.11333 | 5.11333 | 5.12667 | 5.10667
headerMapImplSetInlineMacro/0 | 9.65333 | 9.70667 | 10.0633 | 9.69 | 9.67667 | 9.70667 | 9.68333 | 10.1767 | 9.68 | 9.70333 | 9.69 | 9.69
headerMapImplSetInlineMacro/1 | 9.89667 | 9.71333 | 9.68667 | 9.7 | 9.68333 | 9.69 | 9.68 | 10.1767 | 9.67333 | 9.69 | 9.93667 | 9.68
headerMapImplSetInlineMacro/5 | 9.64667 | 9.71 | 9.68667 | 9.68333 | 9.67667 | 9.68 | 9.68667 | 10.1733 | 9.67 | 9.69 | 9.77333 | 9.68333
headerMapImplSetInlineMacro/10 | 9.64333 | 9.70667 | 9.68667 | 9.70667 | 9.68 | 9.69333 | 10.0767 | 10.2033 | 9.68 | 9.69333 | 9.68667 | 9.68667
headerMapImplSetInlineMacro/50 | 9.66667 | 9.71667 | 9.69333 | 9.68667 | 9.69333 | 9.67333 | 9.67 | 10.2033 | 9.69 | 9.69 | 9.69 | 9.68333
headerMapImplSetInlineInteger/0 | 23.9 | 23.8 | 25.1 | 23.8 | 23.7667 | 23.9333 | 23.8 | 23.8 | 23.7667 | 23.7667 | 23.8 | 23.7667
headerMapImplSetInlineInteger/1 | 23.6667 | 23.8 | 23.9333 | 23.7667 | 23.7667 | 23.9 | 23.9333 | 23.7667 | 23.7667 | 23.7667 | 23.8333 | 23.7333
headerMapImplSetInlineInteger/5 | 23.6667 | 23.8 | 24.0667 | 23.7667 | 23.7667 | 23.9 | 24 | 23.7667 | 23.7667 | 23.7667 | 23.7333 | 23.7333
headerMapImplSetInlineInteger/10 | 23.7 | 23.8333 | 23.7667 | 23.8667 | 23.8667 | 23.9 | 24 | 23.9 | 23.7667 | 23.7667 | 23.7333 | 23.8
headerMapImplSetInlineInteger/50 | 23.6333 | 23.9 | 23.8333 | 23.7333 | 23.9 | 24 | 24 | 25.0667 | 23.7333 | 23.8 | 23.8333 | 23.7667
headerMapImplGetByteSize/0 | 1.57333 | 1.62333 | 1.70333 | 1.62667 | 1.63 | 1.63 | 1.62333 | 1.60333 | 1.64 | 1.64333 | 1.62667 | 1.65333
headerMapImplGetByteSize/1 | 1.56667 | 1.62 | 1.70333 | 1.61667 | 1.63 | 1.64 | 1.61333 | 1.6 | 1.64 | 1.62667 | 1.60667 | 1.63667
headerMapImplGetByteSize/5 | 1.56667 | 1.61 | 1.70333 | 1.64333 | 1.62667 | 1.61667 | 1.62667 | 1.60333 | 1.61667 | 1.63667 | 1.62667 | 1.63
headerMapImplGetByteSize/10 | 1.55 | 1.61667 | 1.70667 | 1.63333 | 1.62667 | 1.63333 | 1.61 | 1.59333 | 1.61667 | 1.63 | 1.62 | 1.60333
headerMapImplGetByteSize/50 | 1.56 | 1.64333 | 1.67667 | 1.60333 | 1.62667 | 1.62333 | 1.60667 | 1.6 | 1.62 | 1.64 | 1.60333 | 1.62667
headerMapImplIterate/0 | 6.76333 | 6.22333 | 6.16667 | 6.21333 | 6.20333 | 6.21667 | 6.22 | 6.22667 | 6.23 | 6.26333 | 6.21 | 6.21
headerMapImplIterate/1 | 7.52 | 7.40667 | 7.41 | 7.45667 | 7.45333 | 7.45 | 7.45 | 7.47667 | 7.45667 | 7.49333 | 7.45333 | 7.45333
headerMapImplIterate/5 | 13.5 | 13.3 | 13.1 | 13.2333 | 13.2333 | 13.3 | 13.3 | 13.2333 | 13.2333 | 13.2667 | 13.3 | 13.2333
headerMapImplIterate/10 | 20.3 | 20.0333 | 19.5 | 19.4333 | 19.4 | 19.4333 | 19.4 | 19.4 | 19.4333 | 19.4 | 19.4667 | 19.4
headerMapImplIterate/50 | 83.7667 | 81.6 | 79.2333 | 79.7333 | 78.8333 | 79.2667 | 80.3 | 79.0333 | 78.7667 | 78.8333 | 79.5 | 79.4667
headerMapImplRemove/0 | 59.4 | 72.4667 | 76.1333 | 45.4 | 45.6333 | 46.1667 | 45.5333 | 45.9667 | 46.1667 | 46.1667 | 45.7667 | 47.8
headerMapImplRemove/1 | 69.6333 | 73.5333 | 77.9 | 76 | 50.1333 | 50.5333 | 50.0333 | 50.4 | 50.4 | 50.3 | 50.1667 | 54
headerMapImplRemove/5 | 106.333 | 72.5333 | 76.6667 | 75.0333 | 76.9 | 76.3333 | 76.2333 | 75.1333 | 66.1667 | 66.2 | 66.0333 | 71.1
headerMapImplRemove/10 | 158 | 72.4667 | 84.3667 | 81.4667 | 84 | 84.7667 | 83.7667 | 83.4 | 85.2667 | 82.8333 | 85.3333 | 85.6667
headerMapImplRemove/50 | 354.667 | 73.2667 | 85.1 | 86.2667 | 82.0333 | 80.4667 | 78.2 | 78.9667 | 78.8333 | 81.9 | 80.7667 | 85.5
headerMapImplRemoveInline/0 | 63.7667 | 77.4333 | 74.5 | 76.0667 | 76.1 | 74.3333 | 79.1333 | 77 | 73.9667 | 76.0667 | 74.1 | 77.9333
headerMapImplRemoveInline/1 | 68.5333 | 77.3667 | 74.5333 | 75.9667 | 76.2 | 75.3 | 79.1 | 76.7 | 74.8667 | 76.3 | 73.8333 | 77.8
headerMapImplRemoveInline/5 | 63.7 | 77.9 | 73.6 | 74.8333 | 76.9 | 74.1 | 79.7 | 77.7333 | 75.5 | 76.4 | 75 | 79.2333
headerMapImplRemoveInline/10 | 63.6333 | 77.3333 | 75.9 | 77.7 | 77.2667 | 75.7 | 80.3333 | 77.4333 | 73.8667 | 73.9333 | 73.9667 | 78
headerMapImplRemoveInline/50 | 64 | 77.1667 | 75 | 76.6667 | 76.6333 | 74.4 | 79.1667 | 77.3333 | 74.3 | 75.3667 | 73.9333 | 78.8667
headerMapImplPopulate | 345.667 | 347.667 | 353.667 | 348 | 348.333 | 353 | 351.667 | 347.333 | 353.333 | 355 | 352 | 355.333
headerMapImplEmulateH1toH2Upgrade/0 | 258.333 | 294 | 289.333 | 290.333 | 290.333 | 291 | 288.667 | 290.667 | 289.667 | 289.333 | 291 | 291.333
headerMapImplEmulateH1toH2Upgrade/1 | 322.333 | 372 | 350.667 | 351 | 351.333 | 351 | 348.667 | 351 | 353.333 | 349 | 351.333 | 350.667
headerMapImplEmulateH1toH2Upgrade/5 | 558.333 | 608 | 589.667 | 589.667 | 588.333 | 588 | 587.333 | 590.333 | 590 | 587.333 | 590.333 | 593.333
headerMapImplEmulateH1toH2Upgrade/10 | 861.333 | 897 | 881.333 | 885.667 | 884 | 886.667 | 878.667 | 883.667 | 885 | 880 | 885.333 | 882.667
headerMapImplEmulateH1toH2Upgrade/50 | 3234.67 | 3261 | 3226.67 | 3234.67 | 3236.33 | 3240.33 | 3225.67 | 3227.33 | 3256 | 3246 | 3250.33 | 3240.33
headerMapImplEmulateH2toH1Upgrade/0 | 80.6667 | 83.3667 | 80.3333 | 80.7 | 80.2667 | 80.7333 | 80.9667 | 82.1667 | 81.4 | 80.2333 | 80.5 | 82.0667
headerMapImplEmulateH2toH1Upgrade/1 | 86.8 | 90.3667 | 85.8333 | 86.7667 | 85.9333 | 86.7667 | 85.9 | 88.0333 | 87.5 | 86.2667 | 86.5667 | 87.9
headerMapImplEmulateH2toH1Upgrade/5 | 113.333 | 117.333 | 109 | 110.333 | 110.333 | 110.667 | 110.333 | 112 | 112.333 | 110.333 | 111 | 112
headerMapImplEmulateH2toH1Upgrade/10 | 154 | 158.333 | 145 | 147.333 | 148 | 148.333 | 147.333 | 149.667 | 148 | 147.667 | 147.667 | 149.333
headerMapImplEmulateH2toH1Upgrade/50 | 417.667 | 434 | 375 | 387 | 390 | 388.667 | 388.333 | 391 | 388 | 387.333 | 390.333 | 393.333

The _manyCountryRoutesLongHeaders_ benchmark results:
Original headermap: 106606 ns
Lazymap (threshold=0): 11424 ns

Summary:
- Closer look at the microbenchmarks suggests that the threshold should be set to 2-3, and then the map should be used.
- The benchmarks emulating _encodeHeaders_ and _decodeHeaders_ suggests some penalty on the encoding emulation compared with the original headermap implementation.
- The route matching benchmark shows ~10x improvement in this case.

adisuissa commented:
@mattklein123 @jmarantz @antoniovicente, please take a look.
This is mainly following @jmarantz earlier design.
I've added some tests to try to emulate possible scenarios.

jmarantz commented on source/common/http/header_map_impl.cc:
```diff
@@ -178,6 +178,50 @@ template <> bool HeaderMapImpl::HeaderList::isPseudoHeader(const LowerCaseString
+  if (maybeMakeMap()) {
+    auto iter = lazy_map_.find(key);
+    if (iter != lazy_map_.end()) {
+      // erase from the map, and all same key entries from the list
```
Begin sentences with upper case, end with a period, here & elsewhere in the PR.

adisuissa replied:
Done.

jmarantz commented on source/common/http/header_map_impl.h:
```diff
@@ -180,59 +188,89 @@ class HeaderMapImpl : NonCopyable {
   class HeaderList : NonCopyable {
   public:
+    using HeaderNodeVector = absl::InlinedVector<HeaderNode, 1>;
+    using HeaderLazyMap = absl::flat_hash_map<absl::string_view, HeaderNodeVector>;
```
it may be worth a few sentences here summarizing perf-tests that show it's better to use a flat_hash_map than a std::map here. I became enamored of std::map for this application, IIRC, because of the limited number of headers that are allowed in Envoy, and the performance curve I observed. There's a crossover point where flat_hash_map is better but it might be that the majority of cases are better off with std::map.

I may be remember wrong though.

adisuissa replied:
I've added the results of the benchmarks when using `std::map`. Although there isn't a clear winner, it seems that for thresholds 2-3 either `absl::flat_hash_map` is better or there isn't much difference.
If there are other things we should try, please let me know.

jmarantz replied:
oh...one other (I think I might have tried this earlier) use a std::multimap rather than a map<string_view, vector<HeaderHeaderNode>.

adisuissa replied:
No problem, I'll try `std::multimap`.

jmarantz replied:
The advantage of multimap and map is that you get fast implementation of prefix-matching, which I think maybe I had added a benchmark for in the old PR.

adisuissa replied:
Added usage of `std::multimap` and `absl::btree_multimap` (updated the comparison comment).
For most of the microbenchmarks there is almost no difference compared to `absl::flat_hash_map`, and in some cases the latter is much better.
For the countries benchmark, `std::mutimap` is about 2.5x slower, and `absl::btree_multimap` is about 1.5x slower compared to `absl::flat_hash_map`.
I'll look at the prefix-matching benchmark you added in the old PR.

mattklein123 replied:
Thanks for working on this, this is awesome. It's surprisingly little code for the benefit so I feel good about the additional complexity we are adding for the benefit.

What is the final plan for how we will deploy this? Should we just make the threshhold runtime configurable? It could default at infinite for now if we want default off for initial testing or just pick a reasonable default based on some microbenchmarks of 5-10? If we are worried about perf we could just latch it once when the header map tables are created at startup? This would allow people to play with different thresholds on real systems or possible even tune for their use case?

My preference would be to get this change into a more final form without all the ifdefs and the runtime loading and then go from there? I don't have a strong opinion on flat_map vs. btree though I would prefer to just pick one and go with it. How do you suggest deciding what to go with?

One other thing to think about is whether we even want to support this functionality on trailers? Certainly some deployments may benefit from it but it seems very rare and I'm not sure if it's worth the extra cycles to initialize and check on every operation, though it's potentially mostly noise so no real strong opinion.

Cool stuff!

/wait

adisuissa replied:
Thanks for the comments!

> Thanks for working on this, this is awesome. It's surprisingly little code for the benefit so I feel good about the additional complexity we are adding for the benefit.
> 
> What is the final plan for how we will deploy this? Should we just make the threshhold runtime configurable? It could default at infinite for now if we want default off for initial testing or just pick a reasonable default based on some microbenchmarks of 5-10? If we are worried about perf we could just latch it once when the header map tables are created at startup? This would allow people to play with different thresholds on real systems or possible even tune for their use case?
> 
I think that making it runtime configurable, and loading it when a header map table is created could be costly. That said, I'll implement this to see how costly it is.
Regarding the latching at startup, what is the advantage of a runtime configuration over a bootstrap configuration value (if the value is only set once)?

> My preference would be to get this change into a more final form without all the ifdefs and the runtime loading and then go from there? I don't have a strong opinion on flat_map vs. btree though I would prefer to just pick one and go with it. How do you suggest deciding what to go with?
> 
Good question. I'd say it depends on the perf results. At the moment I lean towards flat_map, but if there's a use-case that is really useful and benefits from btree, then btree it is.
I'll add the prefix matching benchmark that @jmarantz suggested, and we'll see from there.
If there are other use-cases that we should test, please let me know.

> One other thing to think about is whether we even want to support this functionality on trailers? Certainly some deployments may benefit from it but it seems very rare and I'm not sure if it's worth the extra cycles to initialize and check on every operation, though it's potentially mostly noise so no real strong opinion.
> 
To keep it simple, all maps should be treated equal.
Another option is to have a threshold configuration for each of the map types, allowing the user to customize them to their use-cases.

> Cool stuff!
> 
> /wait

mattklein123 replied:
> Regarding the latching at startup, what is the advantage of a runtime configuration over a bootstrap configuration value (if the value is only set once)?

There isn't much difference. Mostly just ability to tune without a config change. If we do put it bootstrap I would recommend using one of the `Runtime` types to still allow it to be runtime configurable. In general I think most users will not configure this so runtime seems fine to be but I don't feel strongly about it.

> Another option is to have a threshold configuration for each of the map types, allowing the user to customize them to their use-cases.

Agreed on keeping it simple. I like the idea of a different threshold for each map type. This shouldn't be very difficult and would be easy to add.

adisuissa replied:
I've added a runtime feature configuration that is being read and stored upon the first header-map creation.

jmarantz commented on source/common/http/header_map_impl.h:
```diff
@@ -180,59 +188,89 @@ class HeaderMapImpl : NonCopyable {
     std::list<HeaderEntryImpl> headers_;
-    std::list<HeaderEntryImpl>::iterator pseudo_headers_end_;
+    HeaderNode pseudo_headers_end_;
+    HeaderLazyMap lazy_map_;
```
`#if HEADER_MAP_SIZE_THRESHOLD > 0` here?

adisuissa replied:
If it won't be declared here, we'll need to encompass all the rest of the code that accesses the map with a macro.
I prefer not to add it at the moment, to keep the code more readable, but let me know if you think otherwise.

jmarantz replied:
How long do we expect to leave HEADER_MAP_THRESHOLD at 0? If it will be immediately followed up by setting it to something higher in another PR then I agree it's definitely worth not optimizing for.

adisuissa replied:
I tend to agree.
I expect most of the header-maps (specifically request/response) to have at least 2/3 headers, and so far I think that the microbenchmarks/benchmarks suggest that in these cases the threshold should be larger than 0.

jmarantz replied:
I thought 0 was a special-case in your ifdef to say "don't use maps" but I could be mistaken.

But the question of thresholding is what's the cutover point where a map is a win, compared to linear walk over the vector. I thought it might be around 5-10, but it depends on which map and probably usage pattern.

adisuissa replied:
I removed the macro for the threshold, and it is now set as a runtime flag (which is set to uint32::max by default).
The code uses the flat_hash_map by default. If HEADERMAP_TYPE_MULTIMAP is set, uses std::mutlimap, and if both HEADERMAP_TYPE_MULTIMAP and MULTIMAP_ABSL_BTREE are set, uses the btree implementation.

It seems that flat_hash_map with threshold=3 gives a good performance on most benchmarks, and the benchmark that suffers the most is the RemoveInline (~1.16x slower).

jmarantz commented:
your country-code test is a good one. What do we get when we use a std::map for that?

adisuissa commented:
> your country-code test is a good one. What do we get when we use a std::map for that?

That was @antoniovicente's idea (and I agree it is a good test). Compared to `absl::flat_hash_map`, `std::map` is 2.5x slower. I've also updated the comparison comment with this data.

adisuissa commented:
Memory consumption: The original implementation (and the lazy map too) reserves memory for the custom headers in each request/response. Here is a breakdown of the memory usage (no custom defined headers other than the default ones):
- `RequestHeaderMap`:
  - static allocation: 1,010KiB for the Trie.
  - per-request: 460 bytes.
- `ResponseHeaderMap`:
  - static allocation: 546KiB for the Trie.
  - per-response: 258 bytes.

adisuissa commented:
> It'd be great to get this in. What's the next step?

Sorry, forgot to update this PR.
Summary so far:
Updated the lazy-map to use a runtime feature ("envoy.http.headermap.lazy_map_min_size"), and tried to introduce new removePrefix tests (couldn't find the tests you mentioned in previous non-merged PRs).
I haven't seen differences in these tests too, and realized that I need to reimplement removePrefix when using a BTree. Should be done by tomorrow.

adisuissa commented:
I've added a benchmark for removePrefix, and changed the method's implementation when using multimap.
The benchmark creates a headermap with 80 headers that don't some prefix, and repeatedly adds a variable number of headers with the prefix (and then removes them).

Summary of results:
* In this scenario using multimap is always better (up to almost 2x, and it makes sense because its complexity is O(log(n)) +  O(#headers w/ prefix)).
* It depends on the number of headers w/ prefix and headers w/o prefix. Multimap performs much better as the number of headers w/o prefix increases, because removePrefix doesn't waste time iterating over all the headers. If, however, the number of headers w/o prefix isn't large, or if the number of headers w/ prefix is much larger than the ones w/o prefix, then there won't be much difference between the two.
* One thing to keep in mind is that this is a stress test for removePrefix, and in real world scenarios it probably will only be used a handful of times for requests and responses.

Data: 
RemovePrefix tests | Original | std::mmap-Threshold0
-- | -- | --
headerMapImplRemovePrefix/0 | 1123 | 630
headerMapImplRemovePrefix/1 | 1201 | 679
headerMapImplRemovePrefix/5 | 1322 | 759
headerMapImplRemovePrefix/10 | 1442 | 863
headerMapImplRemovePrefix/50 | 2440 | 1869

adisuissa commented:
Thanks, although I realized I have a bug in the lazymap runtime config.
The server first creates a headermap (and initializes the constant threshold), and only then reads the bootstrap config.
I still want to set a single static threshold (to reduce impact on headermap creation time), but to let it be modifiable after the runtime configuration is initialized.

jmarantz commented:
Yeah, races with the 'runtime' system init always worry me (@antoniovicente got bit by this earlier). In this case does it need to be fixed? If the first header-map has the wrong config will that be a functional issue? The nature of the 'runtime' system is that this is for settings the can change while the system is running, right?

Maybe you can just call out in clear comments that the first header-map won't have the expected lazy-map threshold.

adisuissa commented:
I first tried to read from runtime on each headermap creation, but it was too costly (7x penalty on the headermap creation benchmark). The main reason for doing this in construction time was to allow config updates to modify it, while not changing it during the lifetime of a each headermap.
Converting the threshold to a static member solved that problem, but it requires the lazymap a single initialization of the lazymap.
I can make the static threshold mutable, and provide a setter for it which is called after a configuration update, but that's an ugly workaround.

jmarantz commented:
I see...so the problem is that the static value you have gets latched too early. Yeah.

Once you have singletons everything is really ugly and this is I guess a manifestation of that. I think we might need to poke in a new value (perhaps stored as an atomic uint32_t) when runtime is updated?

adisuissa commented:
I'm not sure if it makes much difference whether it would be atomic or static in this case.

Actually, thinking about it again, the issue I'm trying to avoid is changing the threshold of an already initialized instance of a headermap (if the threshold changes in the middle, it might cause some inconsistencies).
Looking at the code again, it should be ok, as `maybeMakeMap` only looks at the threshold in case the map is empty, and there are 2 cases which cause this:
1. the lazy map wasn't used at all because the threshold wasn't reached.
2. all headers were removed from the headermap.
In both cases a modification to the threshold should not cause inconsistencies, so it should be ok.

I'll add a comment to the code to clarify this.

jmarantz commented:
we'd need it to be atomic if we were going to allow writing it after it was initialized.

we need to write-after-init if we the lazy-init latches onto a runtime value that has not been loaded from bootstrap yet.

adisuissa commented:
Yes, you are correct, my bad.
It needs to be atomic.

mattklein123 commented on docs/root/version_history/current.rst:
```diff
@@ -74,6 +74,7 @@ New Features
 * http: added support for :ref:`%DOWNSTREAM_PEER_FINGERPRINT_1% <config_http_conn_man_headers_custom_request_headers>` as custom header.
 * http: added :ref:`allow_chunked_length <envoy_v3_api_field_config.core.v3.Http1ProtocolOptions.allow_chunked_length>` configuration option for HTTP/1 codec to allow processing requests/responses with both Content-Length and Transfer-Encoding: chunked headers. If such message is served and option is enabled - per RFC Content-Length is ignored and removed.
 * http: introduced new HTTP/1 and HTTP/2 codec implementations that will remove the use of exceptions for control flow due to high risk factors and instead use error statuses. The old behavior is used by default, but the new codecs can be enabled for testing by setting the runtime feature `envoy.reloadable_features.new_codec_behavior` to true. The new codecs will be in development for one month,  ...
+* http: modified the HTTP header-map data-structure to use an underlying dictionary and a list (no change to the header-map API). By default the use of a dictionary is disabled. It can be enabled by setting the `envoy.http.headermap.lazy_map_min_size` runtime feature to a non-negative number which defines the minimal number of headers in a request/response/trailers required for using a dictionary ...
```
Presumably eventually we want to actually configure this by default to some value? Should we rephrase slightly that this is in place for initial testing, and possibly recommend a threshold? Then we will follow up later with a change to lock in a default once we have done some testing?

adisuissa replied:
I've updated the description, adding a suggested value.
I agree that once we find a good threshold (after more testing in real world workloads), we can set it and notify about it in future versions.

mattklein123 commented on source/common/http/header_map_impl.cc:
```diff
@@ -178,6 +179,78 @@ template <> bool HeaderMapImpl::HeaderList::isPseudoHeader(const LowerCaseString
+// Loading the threshold for lazy map once to avoid changing it while an
+// header-map object is active, and to avoid costly calls to Runtime::getInteger.
+uint32_t lazyMapMinSize() {
+  CONSTRUCT_ON_FIRST_USE(
+      uint32_t, static_cast<uint32_t>(Runtime::getInteger("envoy.http.headermap.lazy_map_min_size",
+                                                          std::numeric_limits<uint32_t>::max())));
```
As you already discovered, this is going to be first done in `HeaderMapImplUtility::getAllHeaderMapImplInfo()` and there is no Runtime at that point, so we have to somehow late update this. This is actually really ugly since not only is runtime initialized later, but it can be continuously updated either via filesystem, RTDS, etc. Here is my concrete suggestion:
1) Confirm this can be updated at runtime without issue (I think you already confirmed this but we should add tests and comments, etc.)
2) This is more work (sorry), but we have had multiple cases in which we actually need a callback that gets notified when runtime changes. Basically, I would build a system in which you can register a callback for when the snapshot gets updated, either via RTDS, filesystem. whatever.
3) Install a callback that hits the atomic if the key is set/changed.

(2) will be a good improvement for everyone, but will also enable your use case.

I would also potentially recommend consolidating this inside `HeaderMapImplUtility` so we have a single place to audit all of this type of init logic, but up to you.

adisuissa replied:
After discussing this with Josh and Alyssa offline, I've decided to read the value during the headermap instantiation.
This increases the headermap creation benchmark time from ~15ns to ~23ns when the runtime is not initialized, and to ~88ns when it is.
As for each request/response we only create a handful of headermaps, the impact should be reduced.

That said, I can see the need for (2), and if it's ok, I would like to work on that in a different PR.
What do you think?

mattklein123 replied:
Sure I think it's fine to do it on every creation.

adisuissa replied:
Done.

mattklein123 commented on source/common/http/header_map_impl.cc:
```diff
@@ -178,6 +179,78 @@ template <> bool HeaderMapImpl::HeaderList::isPseudoHeader(const LowerCaseString
+    }
+    // Add all entries from the list into the map.
+    for (auto node = headers_.begin(); node != headers_.end(); ++node) {
+#ifdef HEADERMAP_TYPE_MULTIMAP
```
I would rather not have the ifdefs. Can we pick one and go with it? We can always change later? What's the current thinking on which to go with? Or do you want to make the type runtime configurable also so it can be experimented with at runtime?

adisuissa replied:
From the tests I've got so far, it seems that flat_hash_map, with a threshold of 3 is a good option.
I'd like to finalize the results once we agree on the instantiation mechanism, double check them, and then remove the ifdefs.

jmarantz replied:
sgtm

mattklein123 replied:
Yes this SGTM. Constructor is fine, so let's run the numbers again and pick the one to use?

/wait

adisuissa replied:
Thanks! I'll run the tests, update the results and push the modified PR.

adisuissa commented:
Updated the comment with the latest results, and removed the btree and multimap.
@jmarantz @mattklein123 ready for review, thanks!

mattklein123 commented on source/common/http/header_map_impl.h:
```diff
@@ -212,27 +228,58 @@ class HeaderMapImpl : NonCopyable {
+            if (values_vec.size() == 1) {
+              lazy_map_.erase(entry.key().getStringView());
+            } else {
+              values_vec.erase(std::remove_if(values_vec.begin(), values_vec.end(),
+                                              [&](HeaderNode it) { return it == entry.entry_; }),
+                               values_vec.end());
```
Can you add some comments here? I've read this a few times and I'm not completely clear what is going on inside here

adisuissa replied:
Simplified a bit by splitting the erase and the remove_if calls, and added some comments.

mattklein123 commented on source/common/http/header_map_impl.cc:
```diff
@@ -446,12 +488,23 @@ HeaderEntry* HeaderMapImpl::getExisting(const LowerCaseString& key) {
   }
 
+  // If the lazy map is used, search it instead of iterating the headers list.
+  // TODO(adisuissa): Depending on the key length, it might be faster to do the
+  // trie lookup first for the common O(1) headers.
```
Don't we do that above?

adisuissa replied:
Removed.

mattklein123 reviewed (changes requested):
Thanks this is great work. Just a few small comments. Very excited to see this land.

/wait

mattklein123 commented:
Thanks LGTM. Going to to mark this exempt from stale bot until we land this.

mattklein123 commented:
Thanks for your patience. Can you merge main and then we can do a final pass on this and merge?

/wait

adisuissa commented:
@mattklein123 latest main is now merged, PTAL.

adisuissa commented:
/retest
(failed accessing homebrew's site)

adisuissa commented on bazel/repositories.bzl:
```diff
@@ -595,6 +595,11 @@ def _com_google_absl():
+    native.bind(
+        name = "abseil_btree",
+        actual = "@com_google_absl//absl/container:btree",
+    )
```
Done

mattklein123 commented on docs/root/version_history/current.rst:
```diff
@@ -74,6 +74,7 @@ New Features
 * http: added support for :ref:`%DOWNSTREAM_PEER_FINGERPRINT_1% <config_http_conn_man_headers_custom_request_headers>` as custom header.
 * http: added :ref:`allow_chunked_length <envoy_v3_api_field_config.core.v3.Http1ProtocolOptions.allow_chunked_length>` configuration option for HTTP/1 codec to allow processing requests/responses with both Content-Length and Transfer-Encoding: chunked headers. If such message is served and option is enabled - per RFC Content-Length is ignored and removed.
 * http: introduced new HTTP/1 and HTTP/2 codec implementations that will remove the use of exceptions for control flow due to high risk factors and instead use error statuses. The old behavior is used by default, but the new codecs can be enabled for testing by setting the runtime feature `envoy.reloadable_features.new_codec_behavior` to true. The new codecs will be in development for one month,  ...
+* http: modified the HTTP header-map data-structure to use an underlying dictionary and a list (no change to the header-map API). To conform with previous versions, the use of a dictionary is currently disabled. It can be enabled by setting the `envoy.http.headermap.lazy_map_min_size` runtime feature to a non-negative number which defines the minimal number of headers in a request/response/traile ...
```
Ping on this. Can you open a tracking issue to eventually setting a sensible default for the map threshold?

adisuissa replied:
Tracking issue added: #13333.

mattklein123 commented on source/common/http/header_map_impl.h:
```diff
@@ -213,27 +229,74 @@ class HeaderMapImpl : NonCopyable {
+              // map.
+              lazy_map_.erase(entry.key().getStringView());
+            } else {
+              // Otherwise, find the exact element that is removed from headers_ (entry) in the
+              // vector, and remove it. This keeps an invariant that only the element that is
+              // currently removed from the headers list will also be removed from the map.
+              // Note: it is possible to remove all elements in the vector that satisfy the
+              // predicate during this iteration, but it will complicate the code by requiring
+              // handling of missing values in the lazy map.
+
+              // The call to std::remove_if removes the elements that satisfy the predicate and
+              // shifts the vector elements, but does not resize the vector. The call to erase that
+              // follows erases the unneeded cells (from remove_pos to the end) and modifies the
+              // vector's size.
```
Just to make sure I understand: In the case where the remove_if ends up removing all the values from the vector, we will eventually hit the size == 1 case above and remove the map entry entirely, right? I think this makes sense but I had to think about this for a bit. Can you add a bit of comment about this also?

adisuissa replied:
Yes, you are correct. I've added a comment to describe this case.

mattklein123 reviewed (changes requested):
Thanks, fantastic work. Just 2 small comments. Thank you! @jmarantz any further comments?

/wait

mattklein123 reviewed (approved):
Nice!!!
